### PR TITLE
[Payment request] add PaymentRequest pay route

### PR DESCRIPTION
**Pull request by Prometee**
| Q               | A
|-----------------|-----
| Branch?         | 2.0
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no/
| Deprecations?   | no
| Related tickets | none
| License         | MIT

- [x] Create a new route to handle PaymentRequest pay rendering via the Shop UI (to also be able to replay a capture or any other action which are not completed already).
- [x] Enhance the PaymentRequestPayReponseProvider to retrieve the existing PaymentRequest if it exists already.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/OrderPay/Action/PaymentRequestPayAction.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+use Sylius\Bundle\ResourceBundle\Controller\RequestConfigurationFactoryInterface;
+use Sylius\Component\Payment\Model\PaymentRequestInterface;
+use Sylius\Component\Payment\Repository\PaymentRequestRepositoryInterface;
+use Sylius\Resource\Metadata\MetadataInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+final class PaymentRequestPayAction
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/OrderPay/Action/PaymentRequestPayAction.php:28]:
```suggestion
/** @experimental */
final class PaymentRequestPayAction
```

File: src/Sylius/Bundle/CoreBundle/OrderPay/Provider/PaymentRequestPayResponseProvider.php [hunk 2]
```diff
@@ -54,24 +54,38 @@ public function getResponse(RequestConfiguration $requestConfiguration, OrderInt
...
+
+        $existingPaymentRequest = $this->paymentRequestRepository->findOneByActionPaymentAndMethod(
+            $action,
+            $payment,
+            $paymentMethod
+        );
+
+        if (null === $existingPaymentRequest || $this->finalizedPaymentRequestChecker->isFinal($existingPaymentRequest)) {
+            $paymentRequest->setPayload($this->defaultPayloadProvider->getPayload($paymentRequest));
+            return $paymentRequest;
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/OrderPay/Provider/PaymentRequestPayResponseProvider.php]:
```suggestion
            $paymentRequest->setPayload($this->defaultPayloadProvider->getPayload($paymentRequest));

            return $paymentRequest;
```

File: src/Sylius/Bundle/PaymentBundle/Doctrine/ORM/PaymentRequestRepository.php [hunk 3]
```diff
@@ -90,4 +92,26 @@ public function findByPaymentIdAndStates(mixed $paymentId, array $states): array
...
     }
+
+    public function findOneByActionPaymentAndMethod(
+        string $action,
+        PaymentInterface $payment,
+        PaymentMethodInterface $paymentMethod,
+    ): ?PaymentRequestInterface {
+        $queryBuilder = $this->createQueryBuilder('o');
+
+        return $queryBuilder
```
Comment (GSadee) [inline src/Sylius/Bundle/PaymentBundle/Doctrine/ORM/PaymentRequestRepository.php]:
```suggestion
        return $this->createQueryBuilder('o')
```

File: src/Sylius/Bundle/PaymentBundle/Resources/config/services/processors.xml [hunk 4]
```diff
@@ -21,5 +21,11 @@
             <argument type="service" id="sylius.normalizer.symfony_request" />
         </service>
         <service id="Sylius\Bundle\PaymentBundle\Processor\NotifyPayloadProcessorInterface" alias="sylius.processor.notify_payload" />
+
+        <service id="sylius.processor.payment_request.http_response" class="Sylius\Bundle\PaymentBundle\Processor\HttpResponseProcessor">
```
Comment (GSadee) [inline src/Sylius/Bundle/PaymentBundle/Resources/config/services/processors.xml]:
```suggestion
        <service id="sylius.processor.http_response" class="Sylius\Bundle\PaymentBundle\Processor\HttpResponseProcessor">
```

File: src/Sylius/Component/Payment/Model/PaymentRequest.php [hunk 5]
```diff
@@ -41,7 +41,7 @@ public function __construct(
 
     public function getId(): ?string
     {
-        return $this->getHash()?->toBinary();
+        return $this->getHash()?->__toString();
```
Comment (NoResponseMate) [inline src/Sylius/Component/Payment/Model/PaymentRequest.php:44]:
```suggestion
        return $this->getHash()?->toString();
```
Pretty much the same thing, but this one's a bit less magic

**Review by NoResponseMate — DISMISSED** [review]

File: src/Sylius/Component/Payment/Model/PaymentRequest.php (hunk 5 above)
Comment (Prometee) [reply src/Sylius/Component/Payment/Model/PaymentRequest.php:44]: @NoResponseMate is there a reason to not use `\Stringable` method?

Comment (NoResponseMate) [reply src/Sylius/Component/Payment/Model/PaymentRequest.php:44]: Not really, it just looks better 🤷

Comment (NoResponseMate) [reply src/Sylius/Component/Payment/Model/PaymentRequest.php:44]: Oh, nevermind then, apparently it does not exist in Symfony 6.4

Comment (Prometee) [reply src/Sylius/Component/Payment/Model/PaymentRequest.php:44]: Ok I revert this ;)

**Review by GSadee — APPROVED** [review]

**Pull request merged**